scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,328 @@
+import heapq
+import itertools
+
+import numpy as np
+
+
+class CubatureRule:
+    def __init__(self, nodes, weights):
+        self.nodes = nodes
+        self.weights = weights
+
+    # returns the dimension of the cubature rule
+    # TODO: need to be more precise about what this means
+    def dimension(self):
+        return self.nodes.shape[-1]
+
+    # returns integral_estimate, error_estimate
+    def estimate(self):
+        pass
+
+
+class CubatureRuleLinearError(CubatureRule):
+    def __init__(self, nodes, weights, error_nodes, error_weights):
+        super().__init__(nodes, weights)
+
+        self.error_nodes = error_nodes
+        self.error_weights = error_weights
+
+    # ranges looks like np.array([[0, 1], [2, 3]]) for
+    # \int^0_1 \int_2^3 f(x, y) dydx
+    # f is assumed to be a function from a function from an array of shape
+    # (eval_points, dim_input) where dim_input is the dimension of the input
+    # and should return an array of shape (eval_points, dim_output) where dim_output
+    # is the dimension of the output, e.g. for f(x, y) = (x + y)^alpha where
+    # alpha is one of [.2, .5, .8] would have dim_input=2 and dim_output=3
+    # TODO: consider better name than ranges, e.g. look at nquad
+    def estimate(self, f, ranges):
```

Comment:
Minor point. I think, but am only ~90% sure, that we want the signature of this to be `def estimate(self, f, a, b)` where `a` and `b` denote the n-dimensional lower and upper points of the cube. I understand that `ranges` comes from `nquad`, but I think in this case it will be much simpler for use to go with `a` and `b`. It also is compatible with what `quad_vec` does.

Reply:
At some point we might want to make a decision about integration regions other than hypercubes ("not supported" is a decision, too). 
Also maybe think about `Bounds` from scipy.optimize: https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html#scipy.optimize.Bounds